### PR TITLE
DOCSP-29504-many-to-one-clusters

### PR DESCRIPTION
**Pull request by jason-price-mongodb**
JIRA

https://jira.mongodb.org/browse/DOCSP-29504

Staging

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/cluster-sync/DOCSP-29504-many-to-one-clusters/release-notes/1.7/#sync-many-source-clusters-with-one-destination-cluster

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/cluster-sync/DOCSP-29504-many-to-one-clusters/quickstart/#sync-many-source-clusters-with-one-destination-cluster

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/cluster-sync/DOCSP-29504-many-to-one-clusters/reference/limitations/#mutiple-simultaneous-sync-limitations

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/cluster-sync/DOCSP-29504-many-to-one-clusters/release-notes/1.7/#set-a-migration-name

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/cluster-sync/DOCSP-29504-many-to-one-clusters/reference/configuration/#mongodb-setting-migrationName

https://preview-mongodbjasonpricemongodb.gatsbyjs.io/cluster-sync/DOCSP-29504-many-to-one-clusters/release-notes/1.7/#set-a-migration-name

### CODE REVIEW

File: source/includes/many-with-one-cluster.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Starting in ``mongosync`` 1.7.0, you can sync multiple source clusters
+simultaneously with a destination cluster. For example, consolidate data
+from many small clusters into one central cluster.
```
Comment (ianf-mongodb) [inline source/includes/many-with-one-cluster.rst:3]:
```suggestion
simultaneously with a destination cluster. For example, you can consolidate data
from many small clusters into one central cluster.
```

File: source/includes/migrationName-description.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Starting in ``mongosync`` 1.7.0, you can set a migration name for a sync
+operation. For example, set a migration name to identify the sync
```
Comment (ianf-mongodb) [inline source/includes/migrationName-description.rst:2]:
```suggestion
operation. For example, you can set a migration name to identify the sync
```

File: source/includes/opts/migrationName.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
+.. reference/configuration.txt
+.. reference/mongosync.txt
+
+.. versionadded:: 1.7.0
+
+Sets a migration name for a sync operation. For example, you can set a
+migration name to identify the sync operations from multiple source
+clusters into one destination cluster.
+
+The ``migrationName`` string can contain up to 44 alphanumeric and
```
Comment (ianf-mongodb) [inline source/includes/opts/migrationName.rst:10]: It looks like migrationName also supports spaces, should we call this out in the description?

```suggestion
The ``migrationName`` string can contain up to 44 alphanumeric, space, and
```

File: source/reference/limitations.txt [hunk 4]
```diff
@@ -64,6 +64,36 @@ General Limitations
...
+
+- Two source clusters S1 and S2.
+- A destination cluster D.
+- Databases ``X`` and ``Y`` on both S1 and S2.
+- Collections ``A``, ``B``, ``C``, and ``D`` on both S1 and S2.
+- You can sync both of these combinations:
+
+  - ``X.A`` and ``Y.C`` on S1 with D.
+  - ``X.B`` and ``Y.D`` on S2 with D.
+
```
Comment (ianf-mongodb) [inline source/reference/limitations.txt:87]: **Suggestion**

I would move the first sentence of the explanation here:

```
If D is initially empty, you can sync S1 with D. ``X.A`` and ``X.B`` are copied from S1 to D.
```

It took me a while to understand the concept here, the reason you cannot migrate the 2nd set of name spaces is because the namespaces were already taken as part of this sync.

File: source/reference/limitations.txt [hunk 5]
```diff
@@ -64,6 +64,36 @@ General Limitations
...
+- You cannot sync both of these combinations:
+
+  - ``X.A`` and ``X.B`` on S1 with D.
+  - ``X.A`` and ``X.B`` on S2 with D.
+    
+  If D is initially empty, you can sync S1 with D. ``X.A`` and
+  ``X.B`` are copied from S1 to D. You cannot sync S2 with D because
+  ``X.A`` and ``X.B`` conflict with the collections already on D.
```
Comment (ianf-mongodb) [inline source/reference/limitations.txt:95]: **Suggestion**

I would remove the first sentence here, that was moved up and update to `namespaces`

You cannot sync S2 with D because ``X.A`` and ``X.B`` conflict with the namespaces already on D.

File: source/reference/limitations.txt [hunk 6]
```diff
@@ -64,6 +64,36 @@ General Limitations
...
+
+A namespace is a ``database_name.collection_name`` combination. You can
+only sync namespaces that don't conflict.
+
+For example, consider this scenario:
+
+- Two source clusters S1 and S2.
+- A destination cluster D.
+- Databases ``X`` and ``Y`` on both S1 and S2.
+- Collections ``A``, ``B``, ``C``, and ``D`` on both S1 and S2.
```
Comment (ianf-mongodb) [inline source/reference/limitations.txt:82]: **Suggestion**

You might consider some longer collection names of a standard database schema such as `orders`, `orderlines`, `orderStatus`, `Product`. It is slightly hard to parse the concept here because the collection and database names are so short and unrelatable.

File: source/includes/opts/migrationName.rst [hunk 7]
```diff
@@ -0,0 +1,18 @@
+.. reference/configuration.txt
+.. reference/mongosync.txt
+
+.. versionadded:: 1.7.0
+
+Sets a migration name for a sync operation. For example, you can set a
```
Comment (ianf-mongodb) [inline source/includes/opts/migrationName.rst:6]: I would remove the for example here, there are 3x "For examples" currently for this one parameter section.

```suggestion
Sets a migration name for a sync operation. You can set a
```

File: source/includes/opts/migrationName.rst [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+migration name to identify the sync operations from multiple source
+clusters into one destination cluster.
+
+The ``migrationName`` string can contain up to 44 alphanumeric and
+underscore characters. ``migrationName`` is appended to the string
+``"mongosync_internal_"``.
+
+For example, if you set ``migrationName`` to ``"Cluster 27000 to cluster
+35000 sync"``, the final string is ``"mongosync_internal_Cluster 27000
+to cluster 35000 sync"``.
```
Comment (ianf-mongodb) [inline source/includes/opts/migrationName.rst:16]:
```suggestion
If you specify ``migrationName`` as ``Cluster 27000 to cluster 35000 sync`` the resulting final string is ``mongosync_internal_Cluster 27000 to cluster 35000 sync``.
```

**Review by ianf-mongodb — COMMENTED** [review]
Hi @jason-price-mongodb, Thanks for picking this up, I left a few comments for consideration.

File: source/includes/many-with-one-cluster.rst (hunk 1 above)
Comment (jason-price-mongodb) [reply source/includes/many-with-one-cluster.rst:3]: Done. Though generally, we don't like to repeat the "you can" twice in the same para.

File: source/includes/migrationName-description.rst (hunk 2 above)
Comment (jason-price-mongodb) [reply source/includes/migrationName-description.rst:2]: Done.

File: source/reference/limitations.txt (hunk 4 above)
Comment (jason-price-mongodb) [reply source/reference/limitations.txt:87]: Done.

File: source/reference/limitations.txt (hunk 5 above)
Comment (jason-price-mongodb) [reply source/reference/limitations.txt:95]: Will do.

File: source/reference/limitations.txt (hunk 6 above)
Comment (jason-price-mongodb) [reply source/reference/limitations.txt:82]: Will do.

File: source/includes/opts/migrationName.rst (hunk 8 above)
Comment (jason-price-mongodb) [reply source/includes/opts/migrationName.rst]: Done.

File: source/includes/opts/migrationName.rst (hunk 7 above)
Comment (jason-price-mongodb) [reply source/includes/opts/migrationName.rst:6]: Removed the other "for example" in the middle. That works a bit better, otherwise the user might think the migration name is the main reason for the name.

File: source/includes/opts/migrationName.rst (hunk 3 above)
Comment (jason-price-mongodb) [reply source/includes/opts/migrationName.rst]: Cannot have spaces. Updated text.

**Comment by jason-price-mongodb** [comment]
@ianf-mongodb please re-review. Cheers.

File: source/reference/limitations.txt [hunk 9]
```diff
@@ -64,6 +64,38 @@ General Limitations
...
 
+.. _c2c-limitations_multiple_sync:
+
+Multiple Source Sync Limitations
+--------------------------------
+
+.. include:: /includes/many-with-one-cluster.rst
+
+A namespace is a ``database_name.collection_name`` combination. You can
+only sync namespaces that don't conflict.
```
Comment (ianf-mongodb) [inline source/reference/limitations.txt:75]: **Praise**

This is a great additional piece of information. My only suggestion would be to add a `..tip::` admonition to draw extra attention to this concept since its pivot to understand what can and cannot be synced.

File: source/release-notes/1.7.txt [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+To set a migration name, see:
+
+- :setting:`migrationName` configuration file setting
+- :option:`--migrationName` command line option
+
+Minimum Supported Version
+-------------------------
+
+In 1.7.0, the minimum supported version of MongoDB is 6.0.8.
```
Comment (ianf-mongodb) [inline source/release-notes/1.7.txt:43]: **Suggestion**
A bit off topic and not for this ticket, but would it be easier to manage all the support version in a single file and include them at the bottom similar to how we do platform support for server:

https://www.mongodb.com/docs/manual/administration/production-notes/#platform-support-matrix

File: source/quickstart.txt [hunk 11]
```diff
@@ -252,4 +254,32 @@ Synchronization Notes
...
+.. _c2c-quickstart-many-with-one:
+
+Sync Multiple Source Clusters Simultaneously with a Destination Cluster
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. include:: /includes/many-with-one-cluster.rst
+
+For limitations, see :ref:`c2c-limitations_multiple_sync`.
+
+The following examples connect source clusters running on port 27000 and
```
Comment (ianf-mongodb) [inline source/quickstart.txt:266]: **Suggestion**

I think the port numbers here should be monospaced, i.e `27000`, `27001` and `35000`.

**Review by ianf-mongodb — APPROVED** [review]
Hi @jason-price-mongodb, thanks for the updates. I left a couple of nits / suggestions for consideration but consider them supplemental, LGTM.

File: source/quickstart.txt (hunk 11 above)
Comment (jason-price-mongodb) [reply source/quickstart.txt]: Done.

File: source/reference/limitations.txt (hunk 9 above)
Comment (jason-price-mongodb) [reply source/reference/limitations.txt:75]: Will leave as is. The tips are shown in distracting boxes.

File: source/release-notes/1.7.txt (hunk 10 above)
Comment (jason-price-mongodb) [reply source/release-notes/1.7.txt:43]: As written, it follows the precedent set for previous C2C release notes. I know there is a project under investigation to consolidate support matrices into one file.

**Review by jason-price-mongodb — COMMENTED** [review]
@ianf-mongodb thanks for the review. Passing on to Craven for ext review. Pinged him on Slack as he isn't in the C2C repo. Cheers.

**Review by suspicious-mango — APPROVED** [review]
LGTM

**Comment by jason-price-mongodb** [comment]
@suspicious-mango please re-review and lmk if the version numbers are all correct. Thanks!

File: source/includes/migrationName-description.rst [hunk 12]
```diff
@@ -0,0 +1,4 @@
+Starting in ``mongosync`` 1.8.0, you can set a migration name for a sync
+operation. For example, you can set a migration name to identify the
+sync operations from multiple source clusters into one destination
```
Comment (suspicious-mango) [inline source/includes/migrationName-description.rst:3]:
```suggestion
operation. For example, you can set a migration name to identify each
sync operation from multiple source clusters into one destination
```

File: source/includes/opts/migrationName.rst [hunk 13]
```diff
@@ -0,0 +1,18 @@
+.. reference/configuration.txt
+.. reference/mongosync.txt
+
+.. versionadded:: 1.8.0
+
+Sets a migration name for a sync operation. For example, you can set a
+migration name to identify the sync operations from multiple source
+clusters into one destination cluster.
```
Comment (suspicious-mango) [inline source/includes/opts/migrationName.rst:8]:
```suggestion
migration name to identify each sync operation from multiple source
clusters into one destination cluster.
```

File: source/includes/migrationName-description.rst (hunk 12 above)
Comment (suspicious-mango) [reply source/includes/migrationName-description.rst:3]: The current phrasing is a bit unclear on whether a single migration name applies to all sync operations into one destination or each sync operation representing different source clusters.

**Review by suspicious-mango — CHANGES_REQUESTED** [review]

File: source/includes/migrationName-description.rst (hunk 12 above)
Comment (jason-price-mongodb) [reply source/includes/migrationName-description.rst]: Done.

File: source/includes/opts/migrationName.rst (hunk 13 above)
Comment (jason-price-mongodb) [reply source/includes/opts/migrationName.rst]: Done.

**Review by jason-price-mongodb — COMMENTED** [review]
Applied comments.

**Comment by jason-price-mongodb** [comment]
@suspicious-mango please re-review. Thanks.

File: source/includes/opts/migrationName.rst [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+.. versionadded:: 1.8.0
+
+Sets a migration name for a sync operation. For example, you can set a
+migration name to identify each sync operation from multiple source
+clusters into one destination cluster.
+
+The ``migrationName`` string can contain up to 44 alphanumeric
+and underscore characters. ``migrationName`` is appended to the string
+``"mongosync_internal_"``.
```
Comment (suspicious-mango) [inline source/includes/opts/migrationName.rst:12]:
```suggestion
``"mongosync_internal_"`` to serve as the migration metadata database name.
```

File: source/includes/opts/migrationName.rst [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+
+If you set ``migrationName`` to
+``"cluster_27000_to_cluster_35000_sync"``, the resulting final string is
```
Comment (suspicious-mango) [inline source/includes/opts/migrationName.rst:15]:
```suggestion
``"cluster_27000_to_cluster_35000_sync"``, the resulting mongosync metadata database name is
```

**Review by suspicious-mango — CHANGES_REQUESTED** [review]
I think it's worth clarifying why we are appending migrationName to "mongosync_internal_".

**Review by jason-price-mongodb — COMMENTED** [review]
@suspicious-mango please re-review. Thanks.

**Comment by jason-price-mongodb** [comment]
@suspicious-mango please re-review once more just in case. Thanks!

**Review by suspicious-mango — APPROVED** [review]
LGTM. Thanks!
